### PR TITLE
Multi array support

### PR DESCRIPTION
**Pull request by shafreenAnfar**

### CODE REVIEW

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java [hunk 1]
```diff
@@ -729,9 +729,7 @@ public BValue visit(ArrayMapAccessExpr arrayMapAccessExpr) {
             throw new BallerinaException("variable '" + arrayVarRefExpr.getVarName() + "' is null");
         }
 
-        Expression indexExpr = arrayMapAccessExpr.getIndexExpr();
-        BValue indexValue = indexExpr.execute(this);
-
+        Expression[] indexExpr = arrayMapAccessExpr.getIndexExpr();
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java]: Can we change the method name to getIndexExprs

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java [hunk 2]
```diff
@@ -1945,21 +1947,25 @@ private void handleArrayType(ArrayMapAccessExpr arrayMapAccessExpr) {
...
             }
 
             // Set type of the arrays access expression
-            BType typeOfArray = ((BArrayType) arrayMapVarRefExpr.getType()).getElementType();
-            arrayMapAccessExpr.setType(typeOfArray);
+            BType expectedType =  arrayMapVarRefExpr.getType();
+            for (int i = 0; i < arrayMapAccessExpr.getIndexExpr().length; i++) {
+                expectedType =  ((BArrayType) expectedType).getElementType();
+            }
+            arrayMapAccessExpr.setType(expectedType);
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: Can you please explain the reason to have the above loop?

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/types/BTypes.java [hunk 3]
```diff
@@ -145,9 +145,16 @@ public static BType resolveType(SimpleTypeName typeName, SymbolScope symbolScope
...
         // We should define the arrays type here.
         if (bType != null) {
+            SimpleTypeName childSimpleType = new SimpleTypeName(typeName.getName(), typeName.getPackagePath(), true);
+            childSimpleType.setPkgPath(typeName.getPackagePath());
+            childSimpleType.setDimensions(typeName.getDimensions() - 1);
+            BTypes.resolveType(childSimpleType, symbolScope, location);
+
             BArrayType bArrayType = new BArrayType(typeName.getSymbolName().getName(),
-                    bType, typeName.getPackagePath(), bType.getSymbolScope());
+                    BTypes.resolveType(childSimpleType, symbolScope, location),
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/types/BTypes.java]: BTypes.resolveType(childSimpleType, symbolScope, location);   

The above is invoked two times in the above code. Can't we store the value and used it later.

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/parser/antlr4/BLangAntlr4Listener.java [hunk 4]
```diff
@@ -469,6 +469,7 @@ public void exitTypeName(BallerinaParser.TypeNameContext ctx) {
             // This is an array type
             SimpleTypeName typeName = typeNameStack.peek();
             typeName.setArrayType(true);
+            typeName.setDimensions((ctx.getChildCount() - 1) / 2);
```
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/parser/antlr4/BLangAntlr4Listener.java]: Updated the code with the aforementioned method.

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/types/BTypes.java (hunk 3 above)
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/types/BTypes.java]: Sorry, it was added accidentally and removed it from the code. Thank you for pointing out.

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/builder/BLangModelBuilder.java [hunk 5]
```diff
@@ -493,16 +494,20 @@ public void createVarRefExpr(NodeLocation location, String varName) {
...
+        int i = 0;
+        while (i < dimensions) {
+            indexExprs[i++] = exprStack.pop();
+        }
+        checkArgExprValidity(location, Arrays.asList(indexExprs));
 
         ArrayMapAccessExpr.ArrayMapAccessExprBuilder builder = new ArrayMapAccessExpr.ArrayMapAccessExprBuilder();
         builder.setVarName(symName);
-        builder.setIndexExpr(indexExpr);
+        builder.setIndexExpr(indexExprs);
```
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/builder/BLangModelBuilder.java]: Update the code with the aforementioned change.

File: modules/ballerina-core/src/test/java/org/ballerinalang/model/arraysofarrays/ArraysOfArraysTest.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
+package org.ballerinalang.model.arraysofarrays;
```
Comment (anupama-pathirage) [inline modules/ballerina-core/src/test/java/org/ballerinalang/model/arraysofarrays/ArraysOfArraysTest.java]: Copyright information is missing here

Comment (shafreenAnfar) [inline modules/ballerina-core/src/test/java/org/ballerinalang/model/arraysofarrays/ArraysOfArraysTest.java]: Updated the code with copy right information. Thanks for pointing out.

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java (hunk 1 above)
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java]: Updated the code with the aforementioned method.

File: modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java (hunk 2 above)
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/util/semantics/SemanticAnalyzer.java]: Here ArrayMapAccessExpr itself does not have a type. The type is taken from VariableRefExpr. VariableRefExpr's type is taken by looking up against the variable name. 

IMO, even if ArrayMapAccessExpr has the type, it won't be helpful. For example let's say,

```
int[][][][] xxxx = [threeDarray];
int i = xxxx[0][0][0][0];
```

In the above case, suppose type of ArrayMapAccessExpr is int[][][][]. Then the element type would be int[][][]. But that is not actually the expected element type on the left hand side. Expected element type should be int. 

So in the above code it is derived by looking at the original type and the number of indexes in ArrayMapAccessExpr expression.

File: modules/ballerina-core/src/test/java/org/ballerinalang/model/arraysofarrays/ArraysOfArraysTest.java (hunk 6 above)
Comment (shafreenAnfar) [inline modules/ballerina-core/src/test/java/org/ballerinalang/model/arraysofarrays/ArraysOfArraysTest.java]: Updated the editor with the copy right profile.

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java [hunk 7]
```diff
@@ -729,9 +729,7 @@ public BValue visit(ArrayMapAccessExpr arrayMapAccessExpr) {
             throw new BallerinaException("variable '" + arrayVarRefExpr.getVarName() + "' is null");
         }
 
-        Expression indexExpr = arrayMapAccessExpr.getIndexExpr();
-        BValue indexValue = indexExpr.execute(this);
-
+        Expression[] indexExpr = arrayMapAccessExpr.getIndexExprs();
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java]: s/indexExpr/indexExprs

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java [hunk 8]
```diff
@@ -1013,12 +1014,18 @@ private void assignValueToArrayMapAccessExpr(BValue rValue, ArrayMapAccessExpr l
         ArrayMapAccessExpr accessExpr = lExpr;
         if (!(accessExpr.getType() == BTypes.typeMap)) {
             BArray arrayVal = (BArray) accessExpr.getRExpr().execute(this);
-            BInteger indexVal = (BInteger) accessExpr.getIndexExpr().execute(this);
+
+            Expression[] exprs = accessExpr.getIndexExprs();
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java]: Lets rename exprs to indexExprs for readability

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java [hunk 9]
```diff
@@ -1180,26 +1187,30 @@ private MemoryLocation getMemoryLocation(Expression expression) {
     private void setUnitValue(BValue rValue, BStruct lExprValue, int memoryLocation,
                               StructFieldAccessExpr fieldExpr) {
 
-        Expression indexExpr;
+        Expression[] exprs;
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java]: Lets rename exprs to indexExprs for readability

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java [hunk 10]
```diff
@@ -1257,22 +1268,28 @@ private BValue getUnitValue(BValue currentVal, StructFieldAccessExpr fieldExpr)
         }
 
         // If the lExprValue value is not a struct arrays/map, then the unit value is same as the struct
-        Expression indexExpr;
+        Expression[] indexExpr;
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java]: s/indexExpr/indexExprs

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/ArrayMapAccessExpr.java [hunk 11]
```diff
@@ -34,18 +34,18 @@
 public class ArrayMapAccessExpr extends UnaryExpression implements ReferenceExpr {
     private String varName;
     private SymbolName symbolName;
-    private Expression indexExpr;
+    private Expression[] indexExpr;
     private boolean isLHSExpr;
 
     private ArrayMapAccessExpr(NodeLocation location, SymbolName symbolName,
-                               Expression arrayVarRefExpr, Expression indexExpr) {
+                               Expression arrayVarRefExpr, Expression[] indexExpr) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/ArrayMapAccessExpr.java]: s/indexExpr/indexExprs

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/ArrayMapAccessExpr.java [hunk 12]
```diff
@@ -34,18 +34,18 @@
...
         super(location, null, arrayVarRefExpr);
         this.symbolName = symbolName;
         this.indexExpr = indexExpr;
     }
 
     private ArrayMapAccessExpr(NodeLocation location, String varName,
-                               Expression arrayVarRefExpr, Expression indexExpr) {
+                               Expression arrayVarRefExpr, Expression[] indexExpr) {
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/ArrayMapAccessExpr.java]: s/indexExpr/indexExprs

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/ArrayMapAccessExpr.java [hunk 13]
```diff
@@ -90,7 +90,7 @@ public BValue execute(NodeExecutor executor) {
         private NodeLocation location;
         private SymbolName varName;
         private Expression arrayMapVarRefExpr;
-        private Expression indexExpr;
+        private Expression[] indexExpr;
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/ArrayMapAccessExpr.java]: s/indexExpr/indexExprs

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/types/BTypes.java [hunk 14]
```diff
@@ -145,9 +145,15 @@ public static BType resolveType(SimpleTypeName typeName, SymbolScope symbolScope
         // If bType is not null, then element type of this arrays type is available.
         // We should define the arrays type here.
         if (bType != null) {
+            SimpleTypeName childSimpleType = new SimpleTypeName(typeName.getName(), typeName.getPackagePath(), true);
+            childSimpleType.setPkgPath(typeName.getPackagePath());
+            childSimpleType.setArrayType(typeName.getDimensions() - 1);
```
Comment (sameera-jayasoma) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/types/BTypes.java]: Check whether typeName.getDimensions() is equal to 1, if so don't create a type name

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java (hunk 7 above)
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java]: Updated accordingly

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/ArrayMapAccessExpr.java (hunk 13 above)
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/ArrayMapAccessExpr.java]: Updated accordingly

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/ArrayMapAccessExpr.java (hunk 11 above)
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/ArrayMapAccessExpr.java]: Updated accordingly

File: modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/ArrayMapAccessExpr.java (hunk 12 above)
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/model/expressions/ArrayMapAccessExpr.java]: Updated accordingly

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java (hunk 10 above)
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java]: Updated accordingly

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java (hunk 9 above)
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java]: Updated accordingly

File: modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java (hunk 8 above)
Comment (shafreenAnfar) [inline modules/ballerina-core/src/main/java/org/ballerinalang/bre/BLangExecutor.java]: Updated accordingly

**Pull request merged**